From b7b2dfe7fbc437d18a913202b6aa3fb0c6e25230 Mon Sep 17 00:00:00 2001
From: Weifeng Voon <weifeng.voon@intel.com>
Date: Sat, 23 Jul 2016 12:12:07 +0800
Subject: [PATCH 33/40] x86/platform/bxt: Enable board file for APL LPSS PCI
 mode

This patch is for registering SPI slave for APL platform: Leaf Hill,
Juniper Hill and Oxbow Hill.

Signed-off-by: Weifeng Voon <weifeng.voon@intel.com>
Signed-off-by: Wan Ahmad Zainie <wan.ahmad.zainie.wan.mohamad@intel.com>
---
 arch/x86/Kconfig                  |   8 +++
 arch/x86/platform/bxt/Makefile    |   1 +
 arch/x86/platform/bxt/apl-board.c | 133 ++++++++++++++++++++++++++++++++++++++
 3 files changed, 142 insertions(+)
 create mode 100644 arch/x86/platform/bxt/apl-board.c

diff --git a/arch/x86/Kconfig b/arch/x86/Kconfig
index 9d709fa..64e7350 100644
--- a/arch/x86/Kconfig
+++ b/arch/x86/Kconfig
@@ -1,4 +1,12 @@
 # SPDX-License-Identifier: GPL-2.0
+#enable SPI node
+config APL_SPI_BRD
+	bool "PCI mode SPI support on APL"
+	depends on X86_INTEL_LPSS
+	---help---
+	  This option is needed if were to use Intel Apollo Lake SPI
+	  in PCI mode.
+
 # Select 32 or 64 bit
 config 64BIT
 	bool "64-bit kernel" if ARCH = "x86"
diff --git a/arch/x86/platform/bxt/Makefile b/arch/x86/platform/bxt/Makefile
index 5990592..ee34212 100644
--- a/arch/x86/platform/bxt/Makefile
+++ b/arch/x86/platform/bxt/Makefile
@@ -1 +1,2 @@
 obj-$(CONFIG_X86_INTEL_SBI_APL) += sbi_apl.o
+obj-$(CONFIG_APL_SPI_BRD) += apl-board.o
diff --git a/arch/x86/platform/bxt/apl-board.c b/arch/x86/platform/bxt/apl-board.c
new file mode 100644
index 0000000..55a2770
--- /dev/null
+++ b/arch/x86/platform/bxt/apl-board.c
@@ -0,0 +1,133 @@
+/*
+ * Copyright (c) 2016, Intel Corporation
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ */
+
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/clk.h>
+#include <linux/clkdev.h>
+#include <linux/clk-provider.h>
+#include <linux/spi/spidev.h>
+#include <linux/spi/spi.h>
+#include <linux/spi/pxa2xx_spi.h>
+#include <linux/pwm.h>
+
+static struct pxa2xx_spi_chip chip_data = {
+	.gpio_cs = -EINVAL,
+	.dma_burst_size = 1,
+};
+
+static struct spi_board_info apl_spi_slaves[] = {
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 1,
+		.chip_select = 0,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 1,
+		.chip_select = 1,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 1,
+		.chip_select = 2,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 2,
+		.chip_select = 0,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 2,
+		.chip_select = 1,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 2,
+		.chip_select = 2,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 3,
+		.chip_select = 0,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 3,
+		.chip_select = 1,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	},
+	{
+		.modalias = "spidev",
+		.max_speed_hz = 50000000,
+		.bus_num = 3,
+		.chip_select = 2,
+		.controller_data = &chip_data,
+		.mode = SPI_MODE_0,
+	}
+};
+
+static int apl_spi_board_setup(void)
+{
+	int ret = -1;
+
+	/* Register the SPI devices */
+	if (!spi_register_board_info
+			(apl_spi_slaves, ARRAY_SIZE(apl_spi_slaves))) {
+		ret = 0;
+		pr_warn("\nfailed to register the SPI slaves...\n");
+	} else {
+		pr_debug("\nsuccessfully registered the SPI slaves...\n");
+	}
+	return ret;
+}
+
+static int __init apl_board_init(void)
+{
+	int ret;
+
+	pr_debug("\nregistering APL SPI devices...\n");
+	ret = apl_spi_board_setup();
+	if (ret)
+		goto exit;
+
+exit:
+	return ret;
+}
+arch_initcall(apl_board_init);
+MODULE_LICENSE("GPL v2");
-- 
1.9.1

